### PR TITLE
Pr paddle integration

### PR DESCRIPTION
**Pull request by sfraczek**
This PR contains fixes and features required for integration with PaddlePaddle framework, small fix to a test and a benchmark for testing aeon's performance with config used for PaddlePaddle's image classification task for resnet50 on imagenet.

Its best to review it commit by commit.

### CODE REVIEW

**Comment by sfraczek** [comment]
@AdrShv @arogowie-intel Should I pass `bgr_to_rgb` from augmentation config to image config like is done for `fixed_aspect_ratio` in src/provider.cpp:156 rather than pass it to `image config`? 
**[EDIT] a change is ready I can push it if you agree**

Should I move normalization to `transformer` rather than doing it in `loader`?

**Comment by arogowie-intel** [comment]
I think that `bgr_to_rgb` should be in image configuration since it is only change of image representation. Normalization in contrast to this changes image data, thus should be in bag of augmentations.

**Comment by sfraczek** [comment]
On the other hand, why is fixed_aspect_ratio given through augmentation I wonder? What is the pattern here?

**Comment by sfraczek** [comment]
@arogowie-intel I have moved normalization to augmentation config

File: src/etl_image.cpp [hunk 1]
```diff
@@ -150,7 +156,11 @@ cv::Mat image::transformer::transform_single_image(shared_ptr<augment::image::pa
     else
         expandedImage = rotatedImage;
 
-    cv::Mat croppedImage = expandedImage(img_xform->cropbox);
+    // TODO(sfraczek): add test for this resize short
+    cv::Mat resizedShortImage;
+    image::resize_short(expandedImage, resizedShortImage, img_xform->resize_short_size);
+
```
Comment (AdrShv) [inline src/etl_image.cpp:168]: Does it mean that we always resize short size ? as I remember we don't need to do that in training mode

File: src/etl_image.cpp [hunk 2]
```diff
@@ -203,51 +234,82 @@ void image::loader::load(const vector<void*>& outlist, shared_ptr<image::decoded
...
+        }
+        else
+        {
+            from_to = {0, 0, 1, 1, 2, 2};
+        }
+    }
+    else
+    {
+        from_to = {0, 0};
+    }
```
Comment (AdrShv) [inline src/etl_image.cpp]: what about other channels size, for instance 4 ?

File: src/etl_image.cpp [hunk 3]
```diff
@@ -203,51 +234,82 @@ void image::loader::load(const vector<void*>& outlist, shared_ptr<image::decoded
...
+                    cv::split(input_image, channels);
+                }
+                else
+                {
+                    channels = {b_roi, g_roi, r_roi};
+                    cv::split(input_image, channels);
+                }
+                // channelwise call
+                image::standardize(channels, m_mean, m_stddev);
             }
```
Comment (AdrShv) [inline src/etl_image.cpp:308]: what if we don't need normalize image ? As I can see, you always call that

File: src/image.cpp [hunk 4]
```diff
@@ -84,6 +84,55 @@ void image::resize(const cv::Mat& input, cv::Mat& output, const cv::Size2i& size
...
+    if (mean.empty())
+        return;
+    // prep for division
+    cv::Scalar s_mean, s_stddev;
+    for (int i = 0; i < mean.size(); i++)
+    {
+        s_stddev[i] = stddev[i] ? 1. / stddev[i] : 1.;
+        s_mean[i]   = mean[i];
+    }
+
```
Comment (AdrShv) [inline src/image.cpp]: you can do same thing without loop during init
```cv::Scalar s_mean(mean[0], mean[1], mean[3])```
However I don't understand how it could work. first of all , Scalar always has size =4, so in the loop you try to read 4th element of mean that can cause segmentation fault

File: src/image.cpp [hunk 5]
```diff
@@ -84,6 +84,55 @@ void image::resize(const cv::Mat& input, cv::Mat& output, const cv::Size2i& size
...
+            subtract(input[i], s_mean[i], input[i]);
+            // divide
+            multiply(input[i], s_stddev[i], input[i]);
+        }
+        else
+        {
+            subtract(input[i], s_mean, input[i]);
+            // divide
+            multiply(input[i], s_stddev, input[i]);
+        }
```
Comment (AdrShv) [inline src/image.cpp:168]: 4th element of s_mean is 0. How could OpenCV  know that you have 3th element vector ?? Does it work ?

File: src/image.cpp [hunk 6]
```diff
@@ -84,6 +84,55 @@ void image::resize(const cv::Mat& input, cv::Mat& output, const cv::Size2i& size
...
+    // prep for division
+    cv::Scalar s_mean, s_stddev;
+    for (int i = 0; i < mean.size(); i++)
+    {
+        s_stddev[i] = stddev[i] ? 1. / stddev[i] : 1.;
+        s_mean[i]   = mean[i];
+    }
+
+    for (int i = 0; i < input.size(); i++)
+    {
```
Comment (AdrShv) [inline src/image.cpp:135]: please use for(auto& input_image: input) version of loop, that allow do code more readable, because you avoid write [i] everywhere

File: src/image.cpp [hunk 7]
```diff
@@ -109,15 +158,7 @@ void image::convert_mix_channels(vector<cv::Mat>& source,
         prepared_source = &tmp_source;
     }
 
-    if (prepared_source->size() == 1 && target.size() == 1)
-    {
-        size_t size = target[0].total() * target[0].elemSize();
-        memcpy(target[0].data, (*prepared_source)[0].data, size);
-    }
-    else
```
Comment (AdrShv) [inline src/image.cpp]: why did you remove that optimization?

**Review by AdrShv — CHANGES_REQUESTED** [review]

File: src/image.cpp [hunk 8]
```diff
@@ -84,6 +84,55 @@ void image::resize(const cv::Mat& input, cv::Mat& output, const cv::Size2i& size
...
+    {
+        s_stddev[i] = stddev[i] ? 1. / stddev[i] : 1.;
+        s_mean[i]   = mean[i];
+    }
+
+    for (int i = 0; i < input.size(); i++)
+    {
+        // divide
+        multiply(input[i], 1. / 255., input[i]);
+
```
Comment (AdrShv) [inline src/image.cpp]: Please add cv::

File: src/etl_image.cpp (hunk 1 above)
Comment (sfraczek) [reply src/etl_image.cpp:168]: If resize_short_size is 0 then it doesn't resize. The check is inside this function. The default value is 0.

File: src/etl_image.cpp (hunk 2 above)
Comment (sfraczek) [reply src/etl_image.cpp]: Its not handled. How should we handle that?  Copy 0->0, 1->1, 2->2, 3->3?

File: src/etl_image.cpp (hunk 3 above)
Comment (sfraczek) [reply src/etl_image.cpp:308]: Inside this function there is one check and there is assumption that the check in line 223 https://github.com/NervanaSystems/aeon/pull/6/files#diff-a4ccfbf4e57f797ce3bbb8864ceaf6cbR223 was also performed. If mean/stddev are not provided then it returns without changing the image.

File: src/image.cpp (hunk 6 above)
Comment (sfraczek) [reply src/image.cpp:135]: I iterate over three arrays with this index in this loop so I think I can't.

File: src/image.cpp (hunk 5 above)
Comment (sfraczek) [reply src/image.cpp:168]: I check that if mean and stddev are used then the length of each must be equal to number of channels in https://github.com/NervanaSystems/aeon/pull/6/files#diff-a4ccfbf4e57f797ce3bbb8864ceaf6cbR223

File: src/image.cpp (hunk 7 above)
Comment (sfraczek) [reply src/image.cpp]: I don't remember. I think I could bring it back if I add a check that from_to is  something in the pattern of 0,0 ;    0,0, 1,1 ;     0,0, 1,1, 2,2;     0,0, 1,1, 2,2, 3,3. 
I don't know if it will work in this case https://github.com/NervanaSystems/aeon/pull/6/files#diff-a4ccfbf4e57f797ce3bbb8864ceaf6cbR310 where the target image is a cropped with cv::Rect matrix. I think the data is not contiguous there?

**Review by sfraczek — COMMENTED** [review]
@AdrShv I've answered your questions

File: src/image.cpp (hunk 7 above)
Comment (sfraczek) [reply src/image.cpp]: I could actually check  strides (step[0] and step[1]) and compare them with width and height of images

File: src/etl_image.cpp (hunk 2 above)
Comment (AdrShv) [reply src/etl_image.cpp]: I would to use loop approach from line 263

File: src/etl_image.cpp (hunk 1 above)
Comment (AdrShv) [reply src/etl_image.cpp:168]: ok, but please pull out this condition from resize_shot function

**Review by AdrShv — CHANGES_REQUESTED** [review]

File: src/etl_image.cpp (hunk 3 above)
Comment (AdrShv) [reply src/etl_image.cpp:308]: ok, but please, pull out that check from standarize function

Comment (sfraczek) [reply src/etl_image.cpp:308]: @AdrShv should I move it outside?

File: src/image.cpp (hunk 7 above)
Comment (sfraczek) [reply src/image.cpp]: I check by isContinuous

File: src/etl_image.cpp (hunk 1 above)
Comment (sfraczek) [reply src/etl_image.cpp:168]: done

File: src/etl_image.cpp (hunk 2 above)
Comment (sfraczek) [reply src/etl_image.cpp]: done

File: src/image.cpp (hunk 8 above)
Comment (sfraczek) [reply src/image.cpp]: done

File: src/image.cpp [hunk 9]
```diff
@@ -84,9 +84,65 @@ void image::resize(const cv::Mat& input, cv::Mat& output, const cv::Size2i& size
...
-void image::convert_mix_channels(vector<cv::Mat>& source,
-                                 vector<cv::Mat>& target,
-                                 vector<int>&     from_to)
+void image::resize_short(const cv::Mat& input, cv::Mat& output, const int target_size)
+{
+    auto percent = static_cast<float>(target_size) / std::min(input.cols, input.rows);
+
+    auto resized_width  = static_cast<int>(std::round(input.cols * percent));
+    auto resized_height = static_cast<int>(std::round(input.rows * percent));
+    cv::resize(input, output, cv::Size2i(resized_width, resized_height), 0, 0, CV_INTER_LINEAR);
```
Comment (AdrShv) [inline src/image.cpp]: this code is almost the same as src/augment_image.cpp : line 193. May be we can create separate function to avoid copy past.

**Review by AdrShv — CHANGES_REQUESTED** [review]

File: src/augment_image.cpp [hunk 10]
```diff
@@ -191,6 +193,16 @@ shared_ptr<augment::image::params> augment::image::param_factory::make_params(
         float      _horizontal_distortion = horizontal_distortion(random);
         cv::Size2f out_shape(output_width * _horizontal_distortion, output_height);
 
+        // TODO(sfraczek): add test for this resize short
+        if (resize_short_size > 0)
+        {
+            auto percent =
+                static_cast<float>(resize_short_size) / std::min(input_width, input_height);
+            auto resized_width  = static_cast<int>(std::round(input_width * percent));
```
Comment (arogowie-intel) [inline src/augment_image.cpp]: Why not to use `nervana::unbiased_round` ?

File: src/augment_image.cpp [hunk 11]
```diff
@@ -191,6 +193,16 @@ shared_ptr<augment::image::params> augment::image::param_factory::make_params(
         float      _horizontal_distortion = horizontal_distortion(random);
         cv::Size2f out_shape(output_width * _horizontal_distortion, output_height);
 
+        // TODO(sfraczek): add test for this resize short
+        if (resize_short_size > 0)
```
Comment (arogowie-intel) [inline src/augment_image.cpp:198]: @AdrShv Why are we doing this only when cropping is enabled?

File: src/etl_image.hpp [hunk 12]
```diff
@@ -170,15 +172,22 @@ class nervana::image::transformer
...
-    shape_type m_stype;
-    uint32_t   m_channels;
+    bool                m_channel_major;
+    bool                m_fixed_aspect_ratio;
+    bool                m_bgr_to_rgb;
+    shape_type          m_stype;
+    uint32_t            m_channels;
+    std::vector<double> m_mean;
+    std::vector<double> m_stddev;
+    std::vector<int>    from_to;
```
Comment (arogowie-intel) [inline src/etl_image.hpp]: Please add `m_` prefix.

File: src/image.cpp [hunk 13]
```diff
@@ -109,7 +165,13 @@ void image::convert_mix_channels(vector<cv::Mat>& source,
         prepared_source = &tmp_source;
     }
 
-    if (prepared_source->size() == 1 && target.size() == 1)
+    bool index  = 0;
+    bool no_mix = std::find_if(from_to.begin(), from_to.end(), [&index](int element) {
```
Comment (arogowie-intel) [inline src/image.cpp]: This is extremely unreadable - what condition did you want to achieve?

**Review by arogowie-intel — CHANGES_REQUESTED** [review]

File: src/image.cpp [hunk 14]
```diff
@@ -105,18 +105,27 @@ void image::resize(const cv::Mat& input,
...
 
-    auto resized_width  = static_cast<int>(std::round(input.cols * percent));
-    auto resized_height = static_cast<int>(std::round(input.rows * percent));
-    cv::resize(input, output, cv::Size2i(resized_width, resized_height), 0, 0, CV_INTER_LINEAR);
+void image::resize_short(const cv::Mat& input, cv::Mat& output, const int target_size)
+{
+    cv::resize(input,
+               output,
+               get_resized_short_size(input.cols, input.rows, target_size),
+               0, 0, CV_INTER_LINEAR);
```
Comment (sfraczek) [inline src/image.cpp]: Maybe we should consider using the interpolation provided by config

File: test/test_loader.cpp [hunk 15]
```diff
@@ -460,7 +460,7 @@ static std::string generate_manifest_file(
     return manifest_name;
 }
 
-TEST(loader, deterministic)
+TEST(DISABLED_loader, deterministic)
```
Comment (sfraczek) [inline test/test_loader.cpp:463]: alternatively to disabling you could hardcode new values that you get after your changes but the question is what has caused the change in values.

File: src/image.cpp [hunk 16]
```diff
@@ -63,30 +90,90 @@ void image::add_padding(cv::Mat& input, int padding, cv::Size2i crop_offset)
     input = paddedImage(cropbox);
 }
 
-void image::resize(const cv::Mat& input, cv::Mat& output, const cv::Size2i& size, bool interpolate)
+void image::resize(const cv::Mat& input,
+                   cv::Mat& output,
+                   const cv::Size2i& size,
+                   const string& inter_method)
 {
```
Comment (AdrShv) [inline src/image.cpp:97]: Why we can't pass to resize cv::InterpolationFlags instead of string ?

Comment (AdrShv) [reply src/image.cpp:97]: ok, now I see why :)

**Review by arogowie-intel — APPROVED** [review]

File: src/augment_image.cpp (hunk 10 above)
Comment (sfraczek) [reply src/augment_image.cpp]: its like in python

**Review by AdrShv — APPROVED** [review]

File: src/augment_image.cpp (hunk 11 above)
Comment (sfraczek) [reply src/augment_image.cpp:198]: I don't know if we should do it for other paths as well. Maybe we should
